src/ascii_str.rs

```diff
@@ -651,40 +700,65 @@ impl<'a> Iterator for Lines<'a> {
             .chars()
             .position(|chr| chr == AsciiChar::LineFeed)
         {
-            let line = if idx > 0 && self.string[idx - 1] == AsciiChar::CarriageReturn {
-                &self.string[..idx - 1]
+            // SAFETY: `idx` is guaranteed to be `1..len`, as we get it from `position` as `0..len` and make sure it's not `0`.
+            let line = if idx > 0
+                && *unsafe { self.string.as_slice().get_unchecked(idx - 1) }
+                    == AsciiChar::CarriageReturn
+            {
+                // SAFETY: As per above, `idx` is guaranteed to be `1..len`
+                unsafe { self.string.as_slice().get_unchecked(..idx - 1).into() }
             } else {
-                &self.string[..idx]
+                // SAFETY: As per above, `idx` is guaranteed to be `0..len`
+                unsafe { self.string.as_slice().get_unchecked(..idx).into() }
             };
-            self.string = &self.string[idx + 1..];
+            // SAFETY: As per above, `idx` is guaranteed to be `0..len`, so at the extreme, slicing `len..` is a valid empty slice.
+            self.string = unsafe { self.string.as_slice().get_unchecked(idx + 1..).into() };
             Some(line)
         } else if self.string.is_empty() {
             None
         } else {
             let line = self.string;
-            self.string = &self.string[..0];
+            // SAFETY: Slicing `..0` is always valid and yields an empty slice
+            self.string = unsafe { self.string.as_slice().get_unchecked(..0).into() };
```

Comment:
It does actually seem smart enough from this test on the playground: [link](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=1b1f4ea3e944bb155c43bc1676253f24).
I'm not sure if it still can make the optimisation in the middle of other code, but a possible workaround is simply to do `self.string = ""`, it should have the same effect. Whichever one you prefer I can go apply

Reply:
I generally try to avoid unsafe blocks. In this case I suspected there is an easy way without unsafe block, but I don't see it. Therefore it's fine the way you did it. I just found this to be more readable, it's basically your suggestion:
```suggestion
            self.string = unsafe { AsciiStr::from_ascii_unchecked(b"") };
```